upstairs/src/lib.rs

```diff
@@ -2140,76 +2108,75 @@ struct BackpressureConfig {
+
+            guest_work: GuestWork {
                 active: HashMap::new(), // GtoS
                 next_gw_id: 1,
                 completed: AllocRingBuffer::new(2048),
-            }),
-
-            iop_tokens: std::sync::Mutex::new(0),
-            bytes_per_iop: None,
-            iop_limit: None,
-
-            bw_tokens: std::sync::Mutex::new(0),
-            bw_limit: None,
-
-            block_size: AtomicU64::new(0),
+            },
 
-            backpressure_us: AtomicU64::new(0),
+            iop_tokens: 0,
+            bw_tokens: 0,
+            backpressure_us: backpressure_us.clone(),
             backpressure_config: BackpressureConfig {
                 bytes_start: 1024u64.pow(3), // Start at 1 GiB
                 bytes_scale: 9.3e-8,         // Delay of 10ms at 2 GiB in-flight
                 queue_start: 0.05,
                 queue_max_delay: Duration::from_millis(5),
             },
-            backpressure_lock: Mutex::new(()),
-            log,
-        }
-    }
+            log: log.clone(),
+        };
+        let guest = Guest {
+            req_tx,
+            limits: limits.clone(),
 
-    pub fn set_iop_limit(&mut self, bytes_per_iop: usize, limit: usize) {
-        self.bytes_per_iop = Some(bytes_per_iop);
-        self.iop_limit = Some(limit);
-    }
+            block_size: AtomicU64::new(0),
 
-    /*
-     * Return IOPs per second
-     */
-    pub fn get_iop_limit(&self) -> Option<usize> {
-        self.iop_limit
+            backpressure_us,
+            backpressure_lock: Mutex::new(()),
+            log,
+        };
+        (guest, io)
     }
 
-    pub fn set_bw_limit(&mut self, bytes_per_second: usize) {
-        self.bw_limit = Some(bytes_per_second);
+    pub fn set_iop_limit(&self, bytes_per_iop: usize, limit: usize) {
+        self.limits.lock().unwrap().iop_limit = Some(IopLimit {
+            bytes_per_iop,
+            iop_limit: limit,
+        });
     }
 
-    pub fn get_bw_limit(&self) -> Option<usize> {
-        self.bw_limit
+    pub fn set_bw_limit(&self, bytes_per_second: usize) {
+        self.limits.lock().unwrap().bw_limit = Some(bytes_per_second);
     }
 
     /*
      * This is used to submit a new BlockOp IO request to Crucible.
      */
     async fn send(&self, op: BlockOp) -> BlockReqWaiter {
         let (brw, res) = BlockReqWaiter::pair();
-        self.reqs.lock().await.push_back(BlockReq { op, res });
-        self.notify.notify_one();
-
+        if let Err(e) = self.req_tx.send(BlockReq { op, res }).await {
+            // This could happen during shutdown, if the up_main task is
+            // destroyed while the Guest is still trying to do work.
+            //
+            // If this happens, then the BlockReqWaiter will immediately return
+            // with CrucibleError::RecvDisconnected (since the oneshot::Sender
+            // will have been dropped into the void).
+            warn!(self.log, "failed to send op to guest: {e}");
```

Comment:
`req_tx` is bounded to 500 requests (value chosen at random).

However, the Crucible upstairs on the other side is constantly processing data from the queue (running `up_main`).  If the queue is temporarily full, we'll just wait here at the `send(..).await`.  To the caller, waiting for things to go into the queue is indistinguishable from waiting for the `BlockReqWaiter`.

Reply:
As long as we can't deadlock ourselves here, where Crucible Upstairs is stuck trying to send something
over the `BlockReqWaiter`.

I bring it up because we have done this to ourselves in the past :)

Reply:
I think we're safe – replying to a `BlockReqWaiter` is a synchronous operation that doesn't require cooperation from the other side, since it's a `tokio::sync::oneshot`.